### PR TITLE
[Executor] Line process to wait for otel exporter

### PR DESCRIPTION
**Pull request by lumoslnt**
# Description

This pull request primarily focuses on improving the process termination and monitoring in the `promptflow` package. Previously the subprocess always run a dead cycle, and it relies on the call `process.terminate()` to kill the subprocess, it caused that the exporter of `opentelemetry` cannot shutdown properly. in this PR, we send the terminate signal to the subprocess and it will break out of the loop, and we will wait it to be ended in the monitor thread. 

The changes are mainly in two files: `src/promptflow/promptflow/executor/_line_execution_process_pool.py` and `src/promptflow/promptflow/executor/_process_manager.py`. The most significant changes include the introduction of a `TERMINATE_SIGNAL` for process termination, modifications to the `_monitor_workers_and_process_tasks_in_thread` and `exec_line_for_queue` functions to handle this signal, and updates to the `end_process` method in the `_process_manager.py` file to wait for the process to exit before termination.

Here are the key changes:

Changes in `src/promptflow/promptflow/executor/_line_execution_process_pool.py`:

* Introduced a `TERMINATE_SIGNAL` constant to signal process termination.
* Modified the `_monitor_workers_and_process_tasks_in_thread` function to put the `TERMINATE_SIGNAL` in the `input_queue` if the process is not in a timeout state.
* Updated the `exec_line_for_queue` function to break the loop and exit the process if the `TERMINATE_SIGNAL` is found in the `input_queue`.

Changes in `src/promptflow/promptflow/executor/_process_manager.py`:

* Updated the `end_process` method to wait for a process to exit after receiving the `TERMINATE_SIGNAL`. If the process is still running after 10 seconds, it is terminated. The method now also handles the `psutil.NoSuchProcess` and `psutil.TimeoutExpired` exceptions. [[1]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258R180-R201) [[2]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258R374-R395)

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 1]
```diff
@@ -364,16 +366,13 @@ def _monitor_workers_and_process_tasks_in_thread(
...
-        # wait 10 seconds to ensure the spans are exported before the process is killed
-        # otherwise there will be some missing spans for lines
-        # TODO (Task 2950080): make line process wait for spans flush
-        from promptflow._trace._start_trace import _is_tracer_provider_configured
-
-        if _is_tracer_provider_configured():
-            time.sleep(10)
+        if not self._is_timeout:
+            input_queue.put(TERMINATE_SIGNAL)
+            time.sleep(1)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:371]: join?

Comment (thy09) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:371]: Maybe we don't need to sleep here, we only need to check the status in end_process

Comment (lumoslnt) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Refined

File: src/promptflow/promptflow/executor/_process_manager.py [hunk 2]
```diff
@@ -177,16 +177,21 @@ def end_process(self, i):
         try:
             pid = self._process_info[i].process_id
             process = psutil.Process(pid)
-            process.terminate()
-            process.wait()
-            self._process_info.pop(i)
+            # Wait for the process to exit.
```
[the remark this answers is not in the document]
Comment (lumoslnt) [reply src/promptflow/promptflow/executor/_process_manager.py:180]: Added.

File: src/promptflow/promptflow/executor/_process_manager.py [hunk 3]
```diff
@@ -177,16 +177,28 @@ def end_process(self, i):
...
-            bulk_logger.warning(f"Process {pid} had been terminated")
+            bulk_logger.warning(f"Process {pid} had been terminated.")
+        except psutil.TimeoutExpired:
+            try:
+                # If the process is still running after waiting 10 seconds, terminate it.
+                process.terminate()
+                process.wait()
+            except Exception as e:
+                bulk_logger.warning(
+                    f"Unexpected error occurred while end process for index {i} and process id {process.pid}. "
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_process_manager.py:192]: The message occurs twice, maybe we can have a variable to store it to avoid duplicate.

**Review by thy09 — APPROVED** [review]

**Review by PeiwenGaoMS — APPROVED** [review]

**Pull request merged**